shared-bindings/board/__init__.h

```diff
@@ -35,21 +35,30 @@
 extern const mp_obj_dict_t board_module_globals;
 STATIC const MP_DEFINE_STR_OBJ(board_module_id_obj, CIRCUITPY_BOARD_ID);
 
-mp_obj_t common_hal_board_get_i2c(void);
-mp_obj_t common_hal_board_create_i2c(void);
+bool common_hal_board_is_i2c(mp_obj_t obj);
+mp_obj_t common_hal_board_get_i2c(const mp_int_t instance);
+mp_obj_t common_hal_board_create_i2c(const mp_int_t instance);
+mp_obj_t board_i2c(size_t n_args, const mp_obj_t *args);
 MP_DECLARE_CONST_FUN_OBJ_0(board_i2c_obj);
 
-mp_obj_t common_hal_board_get_spi(void);
-mp_obj_t common_hal_board_create_spi(void);
+bool common_hal_board_is_spi(mp_obj_t obj);
+mp_obj_t common_hal_board_get_spi(const mp_int_t instance);
+mp_obj_t common_hal_board_create_spi(const mp_int_t instance);
+mp_obj_t board_spi(size_t n_args, const mp_obj_t *args);
 MP_DECLARE_CONST_FUN_OBJ_0(board_spi_obj);
 
-mp_obj_t common_hal_board_get_uart(void);
-mp_obj_t common_hal_board_create_uart(void);
+bool common_hal_board_is_uart(mp_obj_t obj);
+mp_obj_t common_hal_board_get_uart(const mp_int_t instance);
+mp_obj_t common_hal_board_create_uart(const mp_int_t instance);
+mp_obj_t board_uart(size_t n_args, const mp_obj_t *args);
 MP_DECLARE_CONST_FUN_OBJ_0(board_uart_obj);
 
-mp_obj_t board_i2c(void);
-mp_obj_t board_spi(void);
-mp_obj_t board_uart(void);
+#define CIRCUITPY_BOARD_BUS_SINGLETON(name, bus, instance) \
+    STATIC mp_obj_t board_##name(void) { \
+        const mp_obj_t inst = mp_obj_new_int_from_uint(instance); \
```

Comment:
I think that this code will be simpler if you called common_hal functions directly here. The intermediary board_* functions don't provide any value because we no longer need their argument parsing. Leaving it in makes this code confusing as to what it's doing.

You could also have three macros, one for each bus type, instead of calling separate helper functions. `CIRCUITPY_BOARD_UART_SINGLETON` would allow the generated code to be more UART specific.